File: 2025-03-31/Valid_Parentheses.java
package amazon;

import java.util.HashMap;
import java.util.Stack;

public class Valid_Parentheses {
    public static boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        HashMap<Character, Character> map = new HashMap<>();
        map.put(')','(');
        map.put(']','[');
        map.put('}','{');

        for(char c : s.toCharArray()) {
            if(st.isEmpty()) {
                if(map.containsKey(c))
                    return false;
                st.add(c);
            }
            else {
                if(map.containsKey(c) && st.peek() == map.get(c)) {
                    st.pop();
                }
                else {
                    st.add(c);
                }
            }
        }

        return st.isEmpty();
    }
    public static void main(String[] args) {
        System.out.println(isValid("()[]{{}}"));
    }
}

File: 2026-01-26/Valid_Parentheses.py

class Solution:
    def isValid(self, s: str) -> bool:
        if not s or len(s) == 1:
            return False

        map = {
            '}':'{',
            ')':'(',
            ']':'['
        }

        stack = []
        i, n = 0, len(s)
        while i < n:
            while i < n and s[i] in map and len(stack) and stack[-1] == map[s[i]]:
                stack.pop()
                i += 1

            if i == n:
                break

            stack.append(s[i])
            i += 1

        return len(stack) == 0


sol = Solution()
print(sol.isValid("(]"))